server-state-machine.go

```diff
@@ -344,6 +349,31 @@ func (state serverStateStart) Next(hr handshakeMessageReader) (HandshakeState, [
 			return nil, nil, AlertMissingExtension
 		}
 
+		// Figure out if we're going to use a certificate or a raw public key
+		// TODO(rlb): Move this to a negotiation function?
+		if foundExts[ExtensionTypeClientCertType] &&
+			foundExts[ExtensionTypeServerCertType] &&
+			state.Config.AllowRawPublicKeys {
+			foundClient := false
+			for _, t := range clientClientCertType.CertificateTypes {
+				if t == CertificateTypeRawPublicKey {
+					foundClient = true
+					break
+				}
+			}
+
+			foundServer := false
+			for _, t := range clientServerCertType.CertificateTypes {
+				if t == CertificateTypeRawPublicKey {
+					foundServer = true
+					break
+				}
+			}
+
+			// Only support symmetric usage of raw public keys for now
+			connParams.UsingRawPublicKeys = foundClient && foundServer
```

Comment:
BUG?: You should be failing negotiation here if the client asks for RawPublicKeys and not X.509 and you don't allow RawPublicKeys.